eppo_client/bandit.py

```diff
@@ -0,0 +1,267 @@
+):
+    return BanditEvaluation(
+        flag_key,
+        subject_key,
+        subject_attributes,
+        None,
+        None,
+        0.0,
+        0.0,
+        gamma,
+    )
+
+
+@dataclass
+class BanditEvaluator:
+    sharder: Sharder
+    total_shards: int = 10_000
+
+    def evaluate_bandit(
+        self,
+        flag_key: str,
+        subject_key: str,
+        subject_attributes: Attributes,
+        actions_with_contexts: List[ActionContext],
+        bandit_model: BanditModelData,
+    ) -> BanditEvaluation:
+        # handle the edge case that there are no actions
+        if not actions_with_contexts:
+            return null_evaluation(
+                flag_key, subject_key, subject_attributes, bandit_model.gamma
+            )
+
+        action_scores = self.score_actions(
+            subject_attributes, actions_with_contexts, bandit_model
+        )
+
+        action_weights = self.weigh_actions(
+            action_scores,
+            bandit_model.gamma,
+            bandit_model.action_probability_floor,
+        )
+
+        selected_idx, selected_action = self.select_action(
+            flag_key, subject_key, action_weights
+        )
+        return BanditEvaluation(
+            flag_key,
+            subject_key,
+            subject_attributes,
+            selected_action,
+            actions_with_contexts[selected_idx].attributes,
+            action_scores[selected_idx][1],
+            action_weights[selected_idx][1],
+            bandit_model.gamma,
+        )
+
+    def score_actions(
+        self,
+        subject_attributes: Attributes,
+        actions_with_contexts: List[ActionContext],
+        bandit_model: BanditModelData,
+    ) -> List[Tuple[str, float]]:
+        return [
+            (
+                action_context.action_key,
+                (
+                    score_action(
+                        subject_attributes,
+                        action_context.attributes,
+                        bandit_model.coefficients[action_context.action_key],
+                    )
+                    if action_context.action_key in bandit_model.coefficients
+                    else bandit_model.default_action_score
+                ),
+            )
+            for action_context in actions_with_contexts
+        ]
+
+    def weigh_actions(
+        self, action_scores, gamma, probability_floor
+    ) -> List[Tuple[str, float]]:
+        number_of_actions = len(action_scores)
+        best_action, best_score = max(action_scores, key=lambda t: t[1])
+
+        # adjust probability floor for number of actions to control the sum
+        min_probability = probability_floor / number_of_actions
+
+        # weight all but the best action
+        weights = [
+            (
+                action_key,
+                max(
+                    min_probability,
+                    1.0 / (number_of_actions + gamma * (best_score - score)),
+                ),
+            )
```

Comment:
In java, we'd do the extra step of rounding to the shard space (e.g., the closest ten thousandth) to keep weights consistent across programming languages that may have different decimal number implementations under the hood. ([src](https://github.com/Eppo-exp/java-server-sdk/blob/main/src/main/java/com/eppo/sdk/helpers/bandit/FalconBanditModel.java#L83))

Reply:
Hmm I figure they all use the same standard for `doubles` right? and even if not the precision should be much better than 1/10000 -- going to leave as is for now but happy to adjust in a future PR

Reply:
Yeah you're right, all our core languages use IEEE 754 so should be good 🤞